### PR TITLE
Merge torchdim into functorch build

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82454

This moves first-class dimensions, as prototyped in https://github.com/facebookresearch/torchdim
into the functorch build. This makes them availiable for use in PrimTorch more easily.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: functorch/functorch/dim/__init__.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
+import torch
+from typing import Union, Sequence
+import inspect
+import dis
+from .tree_map import tree_flatten, tree_map
+from .wrap_type import wrap_type
+from functorch._C import dim as _C
+_C._patch_tensor_class()
```
Comment (zou3519) [inline functorch/functorch/dim/__init__.py:8]: To check my understanding: the monkey patching for torchdim happens on `import functorch.dim` or `from functorch.dim import ...` and does not happen if someone just calls `import functorch`, right?

**Comment by zou3519** [comment]
Test failures are legit

File: functorch/test/test_dims.py [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+    #     wait=0,
+    #     warmup=1,
+    #     active=2), on_trace_ready=tensorboard_trace_handler(name), with_stack=True) as profiler:
+    #     for _ in range(3):
+    #         lmb()
+    #         profiler.step()
+    print(name, elapsed / r)
+    return elapsed / r
+
+class TestMin(TestCase):
```
Comment (zou3519) [inline functorch/test/test_dims.py:70]: For a test to be runnable via the pytorch testing infrastructure (i.e. run_test.py), it needs to extend from torch.testing._internal.common_utils.TestCase. That might just be a drop-in replacement for unittest.TestCase here

File: functorch/functorch/dim/__init__.py (hunk 1 above)
Comment (zdevito) [reply functorch/functorch/dim/__init__.py:8]: Correct, I was careful that no patching happens unless the dim package is imported. A follow up is to either write a more in-depth `__torch_function__` for getitem/setitem, or to simply merge dim handling into getitem/setitem code.

Comment (zou3519) [reply functorch/functorch/dim/__init__.py:8]: (y) thanks for confirming

**Review by zou3519 — APPROVED** [review]

File: functorch/functorch/csrc/dim/arena.h [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    TensorRef()
+    : impl_(nullptr){}
+    TensorRef(const at::Tensor& t)
+    : impl_(t.unsafeGetTensorImpl()) {}
+    const at::Tensor& operator*() const {
+        return *(at::Tensor*)this;
+    }
+    at::Tensor* operator->() const {
+        return (at::Tensor*)this;
+    }
```
Comment (swolchok) [inline functorch/functorch/csrc/dim/arena.h:205]: these casts have undefined behavior, and because of TBAA I wouldn't be surprised at all if you were actually bitten by this in practice.

see the aliasing rules at https://en.cppreference.com/w/cpp/language/reinterpret_cast and note that C-style casts are interpreted as whichever fancy_cast matches, which will be `reinterpret_cast` in this case https://en.cppreference.com/w/cpp/language/explicit_cast
